### PR TITLE
Added implementations to recompiler.rs

### PR DESCRIPTION
**Pull request by Mou-Ikkai**
Added all 32-bit, 16-bit, and 8-bit register variants of currently implemented 64-bit registers, along with SEVENTY FOUR new OPcodes!

And yes, Ultima, you can merge this. I'm just doing more as I wait for you lol

### CODE REVIEW

File: src/kernel/src/process/module/recompiler.rs [hunk 1]
```diff
@@ -136,28 +136,45 @@ impl<'input> Recompiler<'input> {
             // Transform instruction.
             let (size, end) = match i.code() {
                 Code::Add_r32_rm32 => (self.transform_add_r32_rm32(i), false),
-                Code::Add_rm8_r8 => (self.transform_add_rm8_r8(i), false),
+                Code::Add_r64_rm64 => (self.transform_add_r64_rm64(i), false),
+                Code::Add_rm8_r8 => (self.transform_add_rm8_r8(i), true),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:140]: This one cannot mark and ended. The only kind that can mark as ended is unconditional control transfer without return back.

File: src/kernel/src/process/module/recompiler.rs [hunk 2]
```diff
@@ -136,28 +136,45 @@ impl<'input> Recompiler<'input> {
...
                 Code::Jmp_rm64 => (self.transform_jmp_rm64(i), true),
                 Code::Jmp_rel8_64 | Code::Jmp_rel32_64 => (self.transform_jmp_rel(i), true),
                 Code::Jne_rel8_64 | Code::Jne_rel32_64 => (self.transform_jne_rel(i), false),
-                Code::Jle_rel8_64 | Code::Jne_rel32_64 => (self.transform_jle_rel(i), false),
+                Code::Jns_rel8_64 | Code::Jns_rel32_64 => (self.transform_jns_rel(i), false),
+                Code::Jo_rel8_64 | Code::Jo_rel32_64 => (self.transform_jo_rel(i), true),
                 Code::Js_rel8_64 | Code::Js_rel32_64 => (self.transform_js_rel(i), false),
                 Code::Lea_r32_m => (self.transform_lea32(i), false),
                 Code::Lea_r64_m => (self.transform_lea64(i), false),
+                Code::Mov_r8_imm8 => (self.transform_mov_r8_imm8(i), true),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:177]: Same here.

File: src/kernel/src/process/module/recompiler.rs [hunk 3]
```diff
@@ -136,28 +136,45 @@ impl<'input> Recompiler<'input> {
...
+                Code::Jg_rel8_64 | Code::Jg_rel32_64 => (self.transform_jg_rel(i), false),
+                Code::Jge_rel8_64 | Code::Jge_rel32_64 => (self.transform_jge_rel(i), false),
+                Code::Jl_rel8_64 | Code::Jl_rel32_64 => (self.transform_jl_rel(i), false),
+                Code::Jle_rel8_64 | Code::Jle_rel32_64 => (self.transform_jle_rel(i), false),
                 Code::Jmp_rm64 => (self.transform_jmp_rm64(i), true),
                 Code::Jmp_rel8_64 | Code::Jmp_rel32_64 => (self.transform_jmp_rel(i), true),
                 Code::Jne_rel8_64 | Code::Jne_rel32_64 => (self.transform_jne_rel(i), false),
-                Code::Jle_rel8_64 | Code::Jne_rel32_64 => (self.transform_jle_rel(i), false),
+                Code::Jns_rel8_64 | Code::Jns_rel32_64 => (self.transform_jns_rel(i), false),
+                Code::Jo_rel8_64 | Code::Jo_rel32_64 => (self.transform_jo_rel(i), true),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:173]: Same here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Comment by Mou-Ikkai** [comment]
Final changes I could implement before reaching INVALID error as discussed in our Matrix chat.
`Code: INVALID | Regs: {None, None, None, None} | OP_Kinds: {Register, Register, Register, Register}`
After merge:
@ultimaweapon `I'll try to implement this so your PR should handle that call as the last instruction.`

File: src/kernel/src/process/module/recompiler.rs [hunk 4]
```diff
@@ -823,6 +1579,148 @@ impl<'input> Recompiler<'input> {
...
+            i.len()
+        }
+    }
+
+    fn transform_xchg_rm8_r8(&mut self, i: Instruction) -> usize {
+        // Either operand can be memory for this instruction so we need to check both of it.
+        if i.op0_kind() == OpKind::Memory && i.is_ip_rel_memory_operand() {
+            panic!("XCHG r/m8, r8 with first operand as RIP-relative is not supported yet.");
+        } else if i.op1_kind() == OpKind::Memory && i.is_ip_rel_memory_operand() {
+            panic!("XCHG r8, r/m8 with second operand as RIP-relative is not supported yet.");
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:1717]: We don't need this.

File: src/kernel/src/process/module/recompiler.rs [hunk 5]
```diff
@@ -168,24 +194,62 @@ impl<'input> Recompiler<'input> {
...
+                Code::VEX_Vpxor_xmm_xmm_xmmm128 => (self.transform_vpxor_xmm_xmm_xmmm128(i), false),
+                Code::VEX_Vxorps_xmm_xmm_xmmm128 => (self.transform_vxorps_xmm_xmm_xmmm128(i), false),
+                Code::VEX_Vxorps_ymm_ymm_ymmm256 => (self.transform_vxorps_ymm_ymm_ymmm256(i), false),
+                Code::Wait => (self.preserve(i), false),
+                Code::Xadd_rm32_r32 => (self.transform_xadd_rm32_r32(i), false),
+                Code::Xchg_r32_EAX => (self.transform_xchg_r32_eax(i), false),
+                Code::Xchg_rm8_r8 => (self.transform_xchg_rm8_r8(i), false),
                 Code::Xchg_rm32_r32 => (self.transform_xchg_rm32_r32(i), false),
+                Code::Xor_AL_imm8 => (self.transform_xor_al_imm8(i), false),
+                Code::Xor_EAX_imm32 => (self.transform_xor_eax_imm32(i), false),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:251]: These one can be preserved.

File: src/kernel/src/process/module/recompiler.rs [hunk 6]
```diff
@@ -850,23 +1778,24 @@ impl<'input> Recompiler<'input> {
         i.len()
     }
 
-    /// Get register other than `keep`.
+    // Get register other than `keep`.
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:1781]: `//` and `///` have a different meaning in Rust.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Comment by Mou-Ikkai** [comment]
@ultimaweapon Please for the love of god specify on the imui suggestions because I implemented them to the docs I saw-

**Comment by Mou-Ikkai** [comment]
I hope that change to imul_2 was it, seeing as imul_3 doesn't work due to the other registers being u32, not u8

File: src/kernel/src/process/module/recompiler.rs [hunk 7]
```diff
@@ -192,12 +196,12 @@ impl<'input> Recompiler<'input> {
...
                 Code::Lea_r32_m => (self.transform_lea32(i), false),
                 Code::Lea_r64_m => (self.transform_lea64(i), false),
-                Code::Mov_r8_imm8 => (self.transform_mov_r8_imm8(i), false),
+                Code::Mov_r8_imm8 => (self.preserve(i), false),
                 Code::Mov_r8_rm8 => (self.transform_mov_r8_rm8(i), false),
                 Code::Mov_r32_imm32 => (self.preserve(i), false),
                 Code::Mov_r32_rm32 => (self.transform_mov_r32_rm32(i), false),
                 Code::Mov_r64_rm64 => (self.transform_mov_r64_rm64(i), false),
-                Code::Mov_rm8_imm8 => (self.transform_mov_rm8_imm8(i), false),
+                Code::Mov_rm8_imm8 => (self.preserve(i), false),
```
Comment (ultimaweapon) [inline src/kernel/src/process/module/recompiler.rs:204]: This one cannot be preserved.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Review by ultimaweapon — DISMISSED** [review]

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**